Add options to include OData query parameters

baez90 opened the pull request:
- add optional context to GET requests to allow cancellation
- support $select, $search and $filter where applicable
- make calendar tests optional in case someone has no Office365 and is not able to provide corresponding values

Co-authored-by: Mattes83

Closes #7

TerraTalpi commented:
Hello @baez90, thank you very much for the great work and pull request! I have read your quite intense pull request, the code looks fine to me. Thanks for fixing all my typos 😅 . I have checked out the code locally on my computer and ran all tests, because the secrets are not available in github workflow here. 

I like it very much that the code seems backwards compatible 😄 

Considering the fact that you added a skip option for the calendar tests, I guess you werent able to test them in your environment. I ran all the tests locally on my environment, but i have two errors in the following methods:

1) *GraphClient_test.go* func `TestGraphClient_ListGroupsWithSearchAndFilter` - due to the fact that the query returns `2` results with my parameters, and I have statically written `1` in the source code - dirty programming from my side.
1) *User_test.go* func `TestUser_ListCalendarView`. Error message: `User.ListCalendarView() error = StatusCode is not OK: 400. Body: {"error":{"code":"BadRequest","message":"Query option '$top' was specified more than once, but it must be specified at most once.","innerError":{"date":"2021-08-06T15:11:00","request-id":"c0788b38-xxxx-xxxx-xxxx-c5a133d53c83","client-request-id":"c0788b38--xxxx-xxxx-xxxx-c5a133d53c83"}}} , wantErr false`

I will fix the source code with a new environment variable for the first error and propose the changes as conversations to the required code lines. For the second error I currently dont know, I will investigate this next week and check the query parameters. This error is returned from MS graph API itself from the server.

baez90 commented:
Hi @TerraTalpi I'm excited by your great and positive feedback and glad you like it :smile: 

One of the main goals of us was to keep the code backwards compatible even though I wished passing a context would be mandatory but could still be done with a new method probably in the future and it now at least allows for cancellation.

You are totally right, I don't have a Office365 account hence I was not able to test the calendar stuff unfortunately and I thought it could be neat to share my solution with others facing the same issue.

I'm looking forward to any further kind of feedback and will try to adopt everything as soon as possible :blush: 

I will look into the second error, I might have an idea how this is caused and can be worked around :blush:

TerraTalpi reviewed (changes requested):
I have now added my code proposals for the number of expected results. This fixes my initial code flaw when initially writing the test case. Please add them to your code. The tests now work on my side, please test them on your side as well.

baez90 commented on GraphClient_test.go:
```diff
@@ -143,27 +153,250 @@ func TestGraphClient_ListGroups(t *testing.T) {
+			isGraphClientInitialized := true
+
+			if len(got) != tt.expectedResults {
+				t.Error("GraphClient.ListGroups(): Did not find expected number of results.")
```
Done

baez90 commented on GraphClient_test.go:
```diff
@@ -143,27 +153,250 @@ func TestGraphClient_ListGroups(t *testing.T) {
+				ListWithFilter(fmt.Sprintf("displayName eq '%s'", msGraphExistingGroupDisplayName)),
+				ListWithContext(context.Background()),
+			},
+			expectedResults: 1,
```
I removed the `expectedResults` completely because it seemed to be easier to directly compare with `msGrpahExistingGroupDisplayNameNumRes`

TerraTalpi replied:
Thanks, makes absolute sense.

baez90 commented on GraphClient_test.go:
```diff
@@ -143,27 +153,250 @@ func TestGraphClient_ListGroups(t *testing.T) {
+				ListWithSearch(fmt.Sprintf(`"displayName:%s"`, msGraphExistingGroupDisplayName)),
+				ListWithFilter(fmt.Sprintf("displayName eq '%s'", msGraphExistingGroupDisplayName)),
+			},
+			expectedResults: 1,
```
same as below

baez90 commented on GraphClient_test.go:
```diff
@@ -43,7 +51,8 @@ func TestEnvironmentVariablesPresent(t *testing.T) {
 		t.Fatal("Environment Variable for an existing user in the group named <MSGraphExistingUserPrincipalInGroup> is missing!")
 	}
```
I moved the whole initialization to a `TestMain` method and check everything before the tests are executed this way you don't need to check anything before any test and error handling is a little bit easier.
Also I replaced `strconv.Atoi` with `strconv.ParseUInt()` to avoid negative values directly when parsing :man_shrugging:

baez90 commented on GraphClient_test.go:
```diff
@@ -19,14 +24,17 @@ var (
 	msGraphExistingGroupDisplayName = os.Getenv("MSGraphExistingGroupDisplayName")
```
see below

baez90 commented:
I amended the changes you requested. I hope everything works now also in your setup (please note that I fixed another type in your env variable name `msGraphExistingGroupDisplayNameNumRes` :sweat_smile: )

Let me know if any further change is required :blush: and thank you very much for your patience and feedback!

baez90 commented:
I removed the `noOpOptions` because they weren't necessary any more and added the options to the last missing function. Not sure if I shall add more tests to all functions to validate the parameters are actually valid (should be the case if Microsoft did not something very strange but who knows :grin: )

I hope this is finally it and you can give it a look once more to see if you find anything 'weird'

TerraTalpi commented:
> I removed the `noOpOptions` because they weren't necessary any more and added the options to the last missing function. Not sure if I shall add more tests to all functions to validate the parameters are actually valid (should be the case if Microsoft did not something very strange but who knows 😁 )

I dont understand this. What do you mean by `noOpOptions`? To which function did you add this? You force-pushed the last commit, I may not remember from the previous version you had.

baez90 commented:
I previously had an `noOpOption` implementation for the `*listQueryOptions` to be able to extend the `makeGETAPICall` easily without having to change all public APIs but I removed it later on because it wasn't required any more :blush:
